statistics/handle/unlock_stats_handler.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handle
+
+import (
+	"context"
+
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/statistics/handle/cache"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"go.uber.org/zap"
+)
+
+// RemoveLockedTables remove tables from table locked array.
+// - tids: table ids of which will be unlocked.
+// - pids: partition ids of which will be unlocked.
+// - tables: table names of which will be unlocked.
+// Return the message of skipped tables and error.
+func (h *Handle) RemoveLockedTables(tids []int64, pids []int64, tables []*ast.TableName) (string, error) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnStats)
+	exec := h.mu.ctx.(sqlexec.SQLExecutor)
+
+	_, err := exec.ExecuteInternal(ctx, "BEGIN PESSIMISTIC")
+	if err != nil {
+		return "", err
+	}
+
+	// Load tables to check locked before delete.
+	tableLocked, err := loadLockedTables(ctx, exec, maxChunkSize)
+	if err != nil {
+		return "", err
```

Comment:
It seems like the transaction is not finished. We'd better use `defer finishTransaction(...)` above.

Reply:
Good catch! I totally forgot that.

Reply:
Addressed in [`13cc3fd` (#46478)](https://github.com/pingcap/tidb/pull/46478/commits/13cc3fdf175c81efb78b44dfcc6dd5ed8a1ed5f5)